app/src/main/java/com/kpiroom/bubble/ui/login/LoginActivity.kt

```diff
@@ -1,61 +1,106 @@
+        val anim = LoginAnimation(storage = toggleAuthAnimation)
 
-        val everyEditText = listOf<EditText>(emailEditText, passwordEditText)
+        val everyEditText = listOf<EditText>(emailEditText, passwordEditText, confirmPasswordEditText)
         everyEditText.forEach {
             it.setOnTouchListener { _, _ ->
-                smoothScrollTo(scrollView)
+                if (it.isFocusableInTouchMode) {
+                    anim.smoothScrollToBottom(progressLayout, scrollView).start()
+                }
                 false
             }
         }
-
-        getStartedButton.setOnClickListener {
-            hideKeyboard()
-            progressLayout.progress()
+        changeAuthButton.paintFlags = Paint.UNDERLINE_TEXT_FLAG
+
+        anim.apply {
+         transformEditTextAlpha(confirmPasswordEditText, true)
+         transformAlpha(confirmPasswordBackground, true)
+            onGlobalLayoutListener = ViewTreeObserver.OnGlobalLayoutListener {
+                transformHeight(
+                    confirmPasswordBackground,
+                    forgotPasswordTextView.measuredHeight,
+                    emailEditText.measuredHeight
+                )
+            }
+            confirmPasswordEditText.viewTreeObserver
+                .addOnGlobalLayoutListener(onGlobalLayoutListener)
+
+            transformTextColor(forgotPasswordTextView, 250L, DecelerateInterpolator())
+            transformAlpha(forgotPasswordTextView, false)
+            scaleXY(forgotPasswordTextView)
+            translateY(
+                forgotPasswordTextView,
+                forgotPasswordTextView.y,
+                forgotPasswordTextView.y + confirmPasswordEditText.paddingTop
+            )
         }
+
+
+
+        logic.isNewAccount.observe(this, Observer { isNew ->
+            isNew?.let {
+                if (isNew) {
+                    toggleAuthAnimation.forEach { it.start() }
+                } else {
+                    toggleAuthAnimation.forEach { it.reverse() }
+                }
+            }
+        })
+
+        logic.delayedAction.observe(this, Observer { action ->
+            action?.apply {
+                Handler().postDelayed({ doAction }, delay)
+            }
+        })
     }
 
     override fun dispatchTouchEvent(ev: MotionEvent?): Boolean {
-        if (ev?.setsFocusInView(currentFocus) != true) {
-            hideKeyboard()
+        ev?.let {
+            if (!it.isWithinView(editTextArea) && !it.isWithinView(changeAuthButton)) {
+                hideKeyboard()
+            }
         }
         return super.dispatchTouchEvent(ev)
     }
 
-    private fun MotionEvent.setsFocusInView(view: View?): Boolean {
-        if (view != null &&
-                view.isFocusable &&
-                (action == MotionEvent.ACTION_UP || action == MotionEvent.ACTION_MOVE) &&
-                (view.id in listOf(emailEditText.id, passwordEditText.id))) {
+    private fun MotionEvent.isWithinView(view: View): Boolean {
```

Comment:
: Boolean = (action == MotionEvent.ACTION_UP).filter { it }.let { .... }

Reply:
The correction doesn't seem to be clear enough